streams/src/test/java/org/apache/kafka/streams/integration/KStreamSerDesIntegrationTest.java

```diff
@@ -0,0 +1,491 @@
+                COUNT_TOPIC,
+                AGGREGATION_TOPIC,
+                REDUCE_TOPIC,
+                JOINED_TOPIC);
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @After
+    public void whenShuttingDown() throws Exception {
+        assertTrue(keyIntSerde.configured());
+        assertTrue(valueStringSerde.configured());
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        CLUSTER.deleteAllTopicsAndWait(30_000L);
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    static class MyStringSerializer extends StringSerializer {
+        boolean configured = false;
+        boolean called = false;
+
+        @Override
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            super.configure(configs, isKey);
+            configured = true;
+        }
+
+        @Override
+        public byte[] serialize(final String topic, final String data) {
+            called = true;
+            return super.serialize(topic, data);
+        }
+
+        boolean configured() {
+            return !called || configured;
+        }
+    }
+
+    static class MyStringDeserializer extends StringDeserializer {
+        boolean configured = false;
+        boolean called = false;
+
+        @Override
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            super.configure(configs, isKey);
+            configured = true;
+        }
+
+        @Override
+        public String deserialize(final String topic, final byte[] data) {
+            called = true;
+            return super.deserialize(topic, data);
+        }
+
+        boolean configured() {
+            return !called || configured;
+        }
+    }
+
+    public static class MyStringSerde<K> extends Serdes.WrapperSerde<String> {
+        public MyStringSerde() {
+            super(new MyStringSerializer(), new MyStringDeserializer());
+        }
+
+        @Override
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            super.configure(configs, isKey);
+        }
+
+        public boolean configured() {
+            if (!((MyStringSerializer) this.serializer()).configured()) return false;
+            if (!((MyStringDeserializer) this.deserializer()).configured()) return false;
+            return true;
+        }
+    }
+
+    private static class SimpleProcessor extends AbstractProcessor<String, String> {
+
+        final List<String> valueList;
+
+        SimpleProcessor(final List<String> valueList) {
+            this.valueList = valueList;
+        }
+
+        @Override
+        public void process(final String key, final String value) {
+            valueList.add(value);
+        }
+    }
+    @Test
+    @SuppressWarnings("unchecked")
+    public void shouldSendCorrectResults_NO_OPTIMIZATION() throws Exception {
```

Comment:
In this test, we don't care about correct result if optimization is disable -- in general we want to test Serde-auto-config feature. Test method should be renamed accordingly.

There are multiple ways to structure the test and break it into smaller pieces. Not sure atm what the best way is, but it should be easy to read and understand. Maybe we can have one test for Processor API, with one source, one stateful processor and one sink?

And one more test for the DSL? We should also cover the global state/KTable case.

Reply:
Since it would take some time to add test coverage / properly structure the tests, can the refinement of the integration test be done in a follow-up PR ?

Reply:
I simplified KStreamSerDesIntegrationTest by removing irrelevant test and merging test cases.